File: cli/internal/posttool/hook.go
package posttool

import (
	"bytes"
	"encoding/json"
	"io"
	"os"
	"path"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"harness-data/cli/internal/harness"
	"harness-data/cli/internal/sessionstate"
)

type Output struct {
	HookSpecificOutput HookSpecificOutput `json:"hookSpecificOutput"`
}

type HookSpecificOutput struct {
	HookEventName     string `json:"hookEventName"`
	AdditionalContext string `json:"additionalContext"`
}

type Payload struct {
	SessionID string `json:"session_id"`
	ToolName  string `json:"tool_name"`
	ToolInput struct {
		Command string `json:"command"`
	} `json:"tool_input"`
}

type reportConfig struct {
	Report          string
	RequiredModules []string
}

var reportConfigs = map[string]reportConfig{
	"store-overview": {
		Report:          "store",
		RequiredModules: []string{"overview"},
	},
	"member-overview": {
		Report:          "user",
		RequiredModules: []string{"overview"},
	},
	"financial-overview": {
		Report:          "company",
		RequiredModules: []string{"indicators", "tree", "table"},
	},
}

var reportOrder = []string{"store-overview", "member-overview", "financial-overview"}

func RunClaudeHook(root string, input []byte) (bool, Output, error) {
	var payload Payload
	if err := json.Unmarshal(input, &payload); err != nil {
		return false, Output{}, nil
	}
	if !isShellTool(payload.ToolName) || strings.TrimSpace(payload.ToolInput.Command) == "" {
		return false, Output{}, nil
	}
	sessionID := payload.SessionID
	if sessionID == "" {
		sessionID = os.Getenv("CLAUDE_SESSION_ID")
	}
	if sessionID == "" {
		sessionID = "unknown"
	}
	command := payload.ToolInput.Command
	if !isTemplateStageCommand(command) && !isTemplateInjectionCommand(command) {
		return false, Output{}, nil
	}
	message, outcome, templateRel, err := InjectTemplate(root, sessionID)
	if err != nil {
		return false, Output{}, err
	}
	updated, _ := sessionstate.Load(root, sessionID)
	recordTemplateDiagnostic(root, sessionID, updated, getReportState(updated, "template"), outcome, templateRel)
	return true, buildOutput(message), nil
}

func isShellTool(toolName string) bool {
	return toolName == "Bash" || toolName == "exec_command"
}

func InjectTemplate(root, sessionID string) (message, outcome, templateRel string, err error) {
	state, err := sessionstate.Load(root, sessionID)
	if err != nil {
		return "", "", "", err
	}
	reportState := getReportState(state, "template")
	if state.Mode == "" {
		return "QDM_INJECT_TEMPLATE session state missing. Do not guess a template; run context first.", "missing_session_state", "", sessionstate.Save(root, sessionID, state)
	}
	if state.Mode == sessionstate.ModeFree {
		state.SelectedPlaybook = ""
		state.SelectedTemplate = ""
		state.SelectedPlaybooks = nil
		if err := sessionstate.Save(root, sessionID, state); err != nil {
			return "", "", "", err
		}
		return "QDM_FREE_ANALYSIS current session is free mode. Do not run inject-template; continue free analysis and do not read template files.", "free_mode_no_template", "", nil
	}
	templateRel, validationMessage := selectedTemplatePath(root, state)
	if validationMessage != "" {
		return validationMessage, "template_selection_error", templateRel, sessionstate.Save(root, sessionID, state)
	}
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		return "", "", templateRel, err
	}
	template, err := os.ReadFile(resolver.Resolve(templateRel))
	if err != nil {
		return "QDM_INJECT_TEMPLATE missing " + templateRel + ". Wiki integrity error; do not read any other template.", "missing_template", templateRel, sessionstate.Save(root, sessionID, state)
	}
	state.TemplateInjected = true
	reportState.TemplateInjected = true
	if err := sessionstate.Save(root, sessionID, state); err != nil {
		return "", "", templateRel, err
	}
	return string(stripMarkdownFrontmatter(template)) + finalOutputContract(), "template_injected", templateRel, nil
}

func finalOutputContract() string {
	return "\n\nQDM_DELIVERY_MODE=chat\nQDM_FINAL_OUTPUT_CONTRACT:\n- Use the injected template to organize the final response in the current conversation.\n- Do not write the final result or intermediate analysis result to a file.\n- Only create an export file when the user explicitly asks to export, save, or generate a file.\n"
}

func buildOutput(message string) Output {
	return Output{
		HookSpecificOutput: HookSpecificOutput{
			HookEventName:     "PostToolUse",
			AdditionalContext: message,
		},
	}
}

func shouldRequireTemplateInjection(state sessionstate.File) bool {
	if state.Mode == sessionstate.ModeFree || state.Mode == "" {
		return false
	}
	if state.SelectedTemplate == "" || state.TemplateInjected {
		return false
	}
	return true
}

func templateInjectionRequiredMessage(state sessionstate.File) string {
	var b strings.Builder
	b.WriteString("QDM_TEMPLATE_REQUIRED: data collection command was recorded for a template-backed Harness plan. Do not answer, summarize, calculate a final report, or read template files yet. Run `bin/data-harness-cli inject-template` now. After the PostToolUse hook injects the selected template, use that injected template to produce the final answer.")
	if state.SelectedPlaybook != "" {
		b.WriteString(" selectedPlaybook=")
		b.WriteString(state.SelectedPlaybook)
	}
	if state.SelectedTemplate != "" {
		b.WriteString(" selectedTemplate=")
		b.WriteString(state.SelectedTemplate)
	}
	return b.String()
}

func recordCommandModules(state sessionstate.File, command string) bool {
	handled := false
	for _, reportName := range reportOrder {
		config := reportConfigs[reportName]
		modules := extractReportModules(command, reportName, config)
		if len(modules) == 0 {
			continue
		}
		reportState := getReportState(state, reportName)
		for _, module := range modules {
			if addModule(reportState, module) {
				handled = true
			}
		}
	}
	return handled
}

func extractReportModules(command, reportName string, config reportConfig) []string {
	normalized := normalizeCommand(command)
	lowered := strings.ToLower(normalized)
	if reportName == "financial-overview" {
		modules := extractStandardModules(lowered, config.Report, config.RequiredModules)
		if financialTableCommand(normalized, lowered) && !contains(modules, "table") {
			modules = append(modules, "table")
		}
		return modules
	}
	return extractStandardModules(lowered, config.Report, config.RequiredModules)
}

func extractStandardModules(lowered, report string, required []string) []string {
	pattern := regexp.MustCompile(`\breport\s+` + regexp.QuoteMeta(report) + `\s+(` + strings.Join(required, "|") + `)\b`)
	matches := pattern.FindAllStringSubmatchIndex(lowered, -1)
	var modules []string
	for i, match := range matches {
		module := lowered[match[2]:match[3]]
		segmentEnd := len(lowered)
		if i+1 < len(matches) {
			segmentEnd = matches[i+1][0]
		}
		segment := lowered[match[0]:segmentEnd]
		if module == "tree" && !strings.Contains(segment, "--values") {
			continue
		}
		if !contains(modules, module) {
			modules = append(modules, module)
		}
	}
	return modules
}

func financialTableCommand(normalized, lowered string) bool {
	hasReport := regexp.MustCompile(`\btable\b`).MatchString(lowered) && regexp.MustCompile(`--report\s+company\b`).MatchString(lowered)
	hasIndicator := regexp.MustCompile(`--indicator\s+(ebitda|ebitdacompanyprofit)\b`).MatchString(lowered)
	hasDim := regexp.MustCompile(`(?i)--dim-type\s+(管理区域|manageareaid)(\s|$)`).MatchString(normalized)
	return hasReport && hasIndicator && hasDim
}

func isTemplateInjectionCommand(command string) bool {
	normalized := normalizeCommand(command)
	if !strings.Contains(normalized, "data-harness-cli") || !strings.Contains(normalized, "inject-template") {
		return false
	}
	return regexp.MustCompile(`(^|\s)["']?(\./)?(bin/)?data-harness-cli["']?\s+inject-template(\s|$|["'])`).MatchString(normalized) ||
		regexp.MustCompile(`/data-harness-cli["']?\s+inject-template(\s|$|["'])`).MatchString(normalized)
}

func isTemplateStageCommand(command string) bool {
	normalized := normalizeCommand(command)
	if !strings.Contains(normalized, "data-harness-cli") || !strings.Contains(normalized, "stage template") {
		return false
	}
	return regexp.MustCompile(`(^|\s)["']?(\./)?(bin/)?data-harness-cli["']?\s+stage\s+template(\s|$|["'])`).MatchString(normalized) ||
		regexp.MustCompile(`/data-harness-cli["']?\s+stage\s+template(\s|$|["'])`).MatchString(normalized)
}

func selectedTemplatePath(root string, state sessionstate.File) (string, string) {
	if state.SelectedPlaybook == "" {
		return "", "QDM_INJECT_TEMPLATE no selectedPlaybook in session state. Do not guess a template; continue without template injection."
	}
	if state.SelectedTemplate == "" {
		return "", "QDM_INJECT_TEMPLATE no selectedTemplate in session state. Do not guess a template; continue without template injection."
	}
	if !isAllowedTemplatePath(state.SelectedTemplate) {
		return state.SelectedTemplate, "QDM_INJECT_TEMPLATE template must be templates/... or reports/.../template.md: " + state.SelectedTemplate + "."
	}
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		return state.SelectedTemplate, "QDM_INJECT_TEMPLATE path config error: " + err.Error() + "."
	}
	info, err := os.Stat(resolver.Resolve(state.SelectedTemplate))
	if err != nil || info.IsDir() {
		return state.SelectedTemplate, "QDM_INJECT_TEMPLATE missing " + state.SelectedTemplate + "."
	}
	return state.SelectedTemplate, ""
}

func isAllowedTemplatePath(logical string) bool {
	return strings.HasPrefix(logical, "templates/") ||
		(strings.HasPrefix(logical, "reports/") && path.Base(logical) == "template.md")
}

func stripMarkdownFrontmatter(data []byte) []byte {
	lines := bytes.Split(data, []byte("\n"))
	if len(lines) == 0 || strings.TrimSpace(string(lines[0])) != "---" {
		return data
	}
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(string(lines[i])) == "---" {
			return bytes.Join(lines[i+1:], []byte("\n"))
		}
	}
	return data
}

func normalizeCommand(command string) string {
	return strings.Join(strings.Fields(command), " ")
}

func getReportState(state sessionstate.File, reportName string) *sessionstate.Report {
	if state.Reports == nil {
		state.Reports = map[string]*sessionstate.Report{}
	}
	if reportName == "" {
		reportName = "template"
	}
	current := state.Reports[reportName]
	if current == nil {
		current = &sessionstate.Report{}
		state.Reports[reportName] = current
	}
	if current.RecordedModules == nil {
		current.RecordedModules = []string{}
	}
	return current
}

func addModule(state *sessionstate.Report, module string) bool {
	if contains(state.RecordedModules, module) {
		return false
	}
	state.RecordedModules = append(state.RecordedModules, module)
	return true
}

func recordTemplateDiagnostic(root, sessionID string, session sessionstate.File, state *sessionstate.Report, outcome, templatePath string) {
	if os.Getenv("QDM_HARNESS_DIAG") != "1" {
		return
	}
	resolver, err := harness.NewPathResolver(root)
	if err != nil {
		resolver = harness.PathResolver{Root: root, Paths: harness.PathsConfig{Spec: "spec", Routing: "routing", Playbooks: "playbooks", Templates: "templates"}}
	}
	event := map[string]any{
		"ts":                        time.Now().UTC().Format(time.RFC3339Nano),
		"session_id":                sessionID,
		"event":                     "inject_template",
		"selected_playbook":         session.SelectedPlaybook,
		"selected_playbooks":        session.SelectedPlaybooks,
		"mode":                      session.Mode,
		"composite":                 session.Composite,
		"template_path":             templatePath,
		"template_stats":            pathStats(resolver.Resolve(templatePath)),
		"template_already_injected": state.TemplateInjected,
		"outcome":                   outcome,
	}
	data, err := json.Marshal(event)
	if err != nil {
		return
	}
	dir := sessionstate.DiagnosticsDir(root)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return
	}
	path := filepath.Join(dir, sessionstate.SafeSessionID(sessionID)+".jsonl")
	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
	if err != nil {
		return
	}
	defer file.Close()
	_, _ = file.Write(append(data, '\n'))
}

func pathStats(path string) map[string]any {
	data, err := os.ReadFile(path)
	if err != nil {
		return map[string]any{"path": path, "exists": false, "bytes": 0, "lines": 0}
	}
	return map[string]any{
		"path":   path,
		"exists": true,
		"bytes":  len(data),
		"lines":  countLines(data),
	}
}

func countLines(data []byte) int {
	if len(data) == 0 {
		return 0
	}
	lines := bytes.Count(data, []byte{'\n'})
	if data[len(data)-1] != '\n' {
		lines++
	}
	return lines
}

func contains(values []string, value string) bool {
	for _, item := range values {
		if item == value {
			return true
		}
	}
	return false
}

func ReadHookStdin() ([]byte, error) {
	return io.ReadAll(os.Stdin)
}

func ReportNames() []string {
	names := append([]string(nil), reportOrder...)
	sort.Strings(names)
	return names
}

File: packages/data-harness-cli/src/lib/posttool/hook.js
import { appendFileSync, mkdirSync, readFileSync, statSync } from "node:fs";
import path from "node:path";

import { loadConfig, newPathResolver, normalizeResolverOwners } from "../harness.js";
import { diagnosticsDir, load as loadState, MODE_FREE, safeSessionId, save as saveState } from "../sessionstate.js";
import { isAllowedTemplateSelectionPath } from "../wikis/template-selection.js";

const WORKBUDDY_SESSION_PREFIX = "workbuddy:";

export const REPORT_ORDER = ["store-overview", "member-overview", "financial-overview"];

export const REPORT_CONFIGS = {
  "store-overview": { report: "store", requiredModules: ["overview"] },
  "member-overview": { report: "user", requiredModules: ["overview"] },
  "financial-overview": { report: "company", requiredModules: ["indicators", "tree", "table"] },
};

export function runClaudeHook(root, input, context = null) {
  const payload = parsePayload(input);
  if (!payload || payload.tool_name !== "Bash" || !String(payload.tool_input?.command || "").trim()) {
    return { ok: false, output: null };
  }
  let sessionID = payload.session_id || process.env.CLAUDE_SESSION_ID || "unknown";
  if (context && (!context.workspaceRoot || context.capabilities?.canWriteWorkspace === false) && (isTemplateStageCommand(payload.tool_input.command) || isTemplateInjectionCommand(payload.tool_input.command))) {
    return {
      ok: true,
      output: workspaceRequiredOutput("QDM_WORKSPACE_REQUIRED: template/report operations require workspaceRoot; no project state was written."),
    };
  }
  return runTemplateHook(context || root, payload.tool_input.command, sessionID);
}

export function runWorkBuddyHook(root, input, context = null) {
  const payload = parsePayload(input);
  const command = String(payload?.tool_input?.command || "").trim();
  if (!payload || payload.tool_name !== "Bash" || !command) {
    return { ok: false, output: null };
  }
  const templateCommand = isTemplateStageCommand(command) || isTemplateInjectionCommand(command);
  const metricCommand = isQDMMetricCommand(command);
  if (!templateCommand && !metricCommand) return { ok: false, output: null };
  if (metricCommand && !templateCommand) return { ok: false, output: null };
  if (context && (!context.workspaceRoot || context.capabilities?.canWriteWorkspace === false) && templateCommand) {
    return {
      ok: true,
      output: workBuddySafetyOutput("QDM_WORKSPACE_REQUIRED: template/report operations require workspaceRoot; no project state was written."),
    };
  }

  const sessionID = String(payload.session_id || "").trim();
  if (!sessionID) {
    return {
      ok: true,
      output: workBuddySafetyOutput(
        "QDM_HARNESS_BLOCKED: WorkBuddy did not provide a stable session_id for template injection. Do not guess, read, or use any template; run context in a new WorkBuddy session first.",
      ),
    };
  }
  try {
    loadConfig(root);
  } catch {
    return {
      ok: true,
      output: workBuddySafetyOutput(
        "QDM_HARNESS_UNAVAILABLE: Harness configuration could not be loaded after the template tool call. Do not guess, read, or use a template.",
      ),
    };
  }
  try {
    const result = runTemplateHook(context || root, command, WORKBUDDY_SESSION_PREFIX + sessionID);
    if (!result.ok) return { ok: false, output: null };
    return {
      ok: true,
      output: {
        continue: true,
        hookSpecificOutput: result.output.hookSpecificOutput,
      },
    };
  } catch {
    return {
      ok: true,
      output: workBuddySafetyOutput(
        "QDM_HARNESS_UNAVAILABLE: The selected template could not be injected. Do not read another template, guess its structure, or produce a final report.",
      ),
    };
  }
}

function workBuddySafetyOutput(message) {
  return {
    continue: true,
    systemMessage: message,
    hookSpecificOutput: {
      hookEventName: "PostToolUse",
      additionalContext: message,
    },
  };
}

function workspaceRequiredOutput(message) {
  return {
    hookSpecificOutput: {
      hookEventName: "PostToolUse",
      additionalContext: message,
    },
  };
}

export function runTemplateHook(root, command, sessionID) {
  if (!isTemplateStageCommand(command) && !isTemplateInjectionCommand(command)) {
    return { ok: false, output: null };
  }
  const injected = injectTemplate(root, sessionID);
  const updated = loadState(root, sessionID);
  recordTemplateDiagnostic(root, sessionID, updated, getReportState(updated, "template"), injected.outcome, injected.templateRel);
  return { ok: true, output: buildOutput(injected.message) };
}

export function injectTemplate(root, sessionID) {
  const state = loadState(root, sessionID);
  const reportState = getReportState(state, "template");
  if (!state.mode) {
    saveState(root, sessionID, state);
    return {
      message: "QDM_INJECT_TEMPLATE session state missing. Do not guess a template; run context first.",
      outcome: "missing_session_state",
      templateRel: "",
    };
  }
  if (state.mode === MODE_FREE) {
    state.selected_playbook = "";
    state.selected_template = "";
    state.selected_playbooks = undefined;
    saveState(root, sessionID, state);
    return {
      message:
        "QDM_FREE_ANALYSIS current session is free mode. Do not run inject-template; continue free analysis and do not read template files.",
      outcome: "free_mode_no_template",
      templateRel: "",
    };
  }
  const { templateRel, message: validationMessage } = selectedTemplatePath(root, state);
  if (validationMessage) {
    saveState(root, sessionID, state);
    return { message: validationMessage, outcome: "template_selection_error", templateRel };
  }
  const resolver = newPathResolver(root);
  let template;
  try {
    template = readFileSync(resolver.resolve(templateRel));
  } catch {
    saveState(root, sessionID, state);
    return {
      message: `QDM_INJECT_TEMPLATE missing ${templateRel}. Wiki integrity error; do not read any other template.`,
      outcome: "missing_template",
      templateRel,
    };
  }
  state.template_injected = true;
  reportState.template_injected = true;
  saveState(root, sessionID, state);
  return {
    message: `${stripMarkdownFrontmatter(template.toString("utf8"))}${finalOutputContract()}`,
    outcome: "template_injected",
    templateRel,
  };
}

function finalOutputContract() {
  return "\n\nQDM_DELIVERY_MODE=chat\nQDM_FINAL_OUTPUT_CONTRACT:\n- Use the injected template to organize the final response in the current conversation.\n- Do not write the final result or intermediate analysis result to a file.\n- Only create an export file when the user explicitly asks to export, save, or generate a file.\n";
}

function buildOutput(message) {
  return {
    hookSpecificOutput: {
      hookEventName: "PostToolUse",
      additionalContext: message,
    },
  };
}

export function recordCommandModules(state, command) {
  let handled = false;
  for (const reportName of REPORT_ORDER) {
    const config = REPORT_CONFIGS[reportName];
    const modules = extractReportModules(command, reportName, config);
    if (!modules.length) continue;
    const reportState = getReportState(state, reportName);
    for (const module of modules) {
      if (addModule(reportState, module)) handled = true;
    }
  }
  return handled;
}

function extractReportModules(command, reportName, config) {
  const normalized = normalizeCommand(command);
  const lowered = normalized.toLowerCase();
  if (reportName === "financial-overview") {
    const modules = extractStandardModules(lowered, config.report, config.requiredModules);
    if (financialTableCommand(normalized, lowered) && !modules.includes("table")) modules.push("table");
    return modules;
  }
  return extractStandardModules(lowered, config.report, config.requiredModules);
}

function extractStandardModules(lowered, report, required) {
  const pattern = new RegExp(`\\breport\\s+${escapeRegExp(report)}\\s+(${required.map(escapeRegExp).join("|")})\\b`, "g");
  const matches = [...lowered.matchAll(pattern)];
  const modules = [];
  for (let i = 0; i < matches.length; i++) {
    const match = matches[i];
    const module = match[1];
    const segmentEnd = i + 1 < matches.length ? matches[i + 1].index : lowered.length;
    const segment = lowered.slice(match.index, segmentEnd);
    if (module === "tree" && !segment.includes("--values")) continue;
    if (!modules.includes(module)) modules.push(module);
  }
  return modules;
}

function financialTableCommand(normalized, lowered) {
  const hasReport = /\btable\b/.test(lowered) && /--report\s+company\b/.test(lowered);
  const hasIndicator = /--indicator\s+(ebitda|ebitdacompanyprofit)\b/.test(lowered);
  const hasDim = /--dim-type\s+(管理区域|manageareaid)(\s|$)/i.test(normalized);
  return hasReport && hasIndicator && hasDim;
}

export function isTemplateInjectionCommand(command) {
  return hasHarnessCLICommand(command, ["inject-template"]);
}

export function isTemplateStageCommand(command) {
  return hasHarnessCLICommand(command, ["stage", "template"]);
}

export function isQDMMetricCommand(command) {
  for (const segment of shellCommandSegments(command)) {
    const invocation = unwrapShellCommand(segment);
    if (invocation.length > 0 && isQDMMetricExecutable(invocation[0])) return true;
  }
  return false;
}

function hasHarnessCLICommand(command, expectedArgs) {
  for (const segment of shellCommandSegments(command)) {
    const invocation = unwrapShellCommand(segment);
    if (invocation.length < expectedArgs.length + 1 || !isExecutableNamed(invocation[0], "data-harness-cli")) continue;
    let matched = true;
    for (let i = 0; i < expectedArgs.length; i++) {
      if (invocation[i + 1].toLowerCase() !== expectedArgs[i].toLowerCase()) {
        matched = false;
        break;
      }
    }
    if (matched) return true;
  }
  return false;
}

function isQDMMetricExecutable(word) {
  const normalized = String(word || "").trim().toLowerCase();
  if (
    normalized === "$qdm_metric_cli" ||
    normalized === "$env:qdm_metric_cli" ||
    (normalized.startsWith("${qdm_metric_cli") && normalized.endsWith("}"))
  ) {
    return true;
  }
  return isExecutableNamed(word, "qdm-metric-cli");
}

function isExecutableNamed(word, name) {
  const normalized = String(word || "").trim().replaceAll("\\", "/");
  const base = path.posix.basename(normalized).toLowerCase();
  return base === name || base === `${name}.exe`;
}

export function shellCommandSegments(command) {
  const segments = [];
  let words = [];
  let word = "";
  let quote = "";
  const flushWord = () => {
    if (!word) return;
    words.push(word);
    word = "";
  };
  const flushSegment = () => {
    flushWord();
    if (words.length) segments.push(words);
    words = [];
  };
  const text = String(command || "");
  for (let i = 0; i < text.length; i++) {
    const ch = text[i];
    if (quote) {
      if (ch === quote) {
        quote = "";
        continue;
      }
      if (ch === "\\") {
        if (quote === '"' && i + 1 < text.length && (text[i + 1] === '"' || text[i + 1] === "\\")) {
          word += text[i + 1];
          i += 1;
          continue;
        }
        word += "/";
        continue;
      }
      word += ch;
      continue;
    }
    if (ch === "$" && i + 1 < text.length && text[i + 1] === "{") {
      word += "${";
      i += 2;
      for (; i < text.length; i++) {
        word += text[i];
        if (text[i] === "}") break;
      }
      continue;
    }
    if (ch === "$" && i + 1 < text.length && text[i + 1] === "(") {
      flushSegment();
      i += 1;
      continue;
    }
    if (ch === "'" || ch === '"') {
      quote = ch;
      continue;
    }
    if (ch === "\\") {
      if (i + 1 < text.length && (text[i + 1] === " " || text[i + 1] === "\t" || text[i + 1] === "\n")) {
        word += text[i + 1];
        i += 1;
      } else {
        word += "/";
      }
      continue;
    }
    if (ch === " " || ch === "\t" || ch === "\r") {
      flushWord();
      continue;
    }
    if (ch === "\n" || ch === ";" || ch === "|" || ch === "&" || ch === "(" || ch === ")" || ch === "{" || ch === "}") {
      flushSegment();
      continue;
    }
    word += ch;
  }
  flushSegment();
  return segments;
}

function unwrapShellCommand(words) {
  let i = 0;
  while (i < words.length && (isShellAssignment(words[i]) || isShellControlWord(words[i]))) i += 1;
  while (i < words.length) {
    switch (words[i].toLowerCase()) {
      case "command":
      case "exec":
      case "nohup":
      case "builtin":
      case "time":
        i += 1;
        break;
      case "sudo":
        i += 1;
        while (i < words.length && words[i].startsWith("-")) i += 1;
        break;
      case "env":
        i += 1;
        while (i < words.length && (words[i].startsWith("-") || isShellAssignment(words[i]))) i += 1;
        break;
      default:
        return words.slice(i);
    }
  }
  return [];
}

function isShellControlWord(word) {
  switch (String(word || "").toLowerCase()) {
    case "if":
    case "then":
    case "elif":
    case "else":
    case "while":
    case "until":
    case "do":
      return true;
    default:
      return false;
  }
}

function isShellAssignment(word) {
  const cut = String(word || "").indexOf("=");
  if (cut <= 0) return false;
  const name = word.slice(0, cut);
  for (let i = 0; i < name.length; i++) {
    const code = name.charCodeAt(i);
    const isAlpha = (code >= 65 && code <= 90) || (code >= 97 && code <= 122) || code === 95;
    const isDigit = code >= 48 && code <= 57;
    if (!isAlpha && (i === 0 || !isDigit)) return false;
  }
  return true;
}

function selectedTemplatePath(root, state) {
  if (!state.selected_playbook) {
    return {
      templateRel: "",
      message: "QDM_INJECT_TEMPLATE no selectedPlaybook in session state. Do not guess a template; continue without template injection.",
    };
  }
  if (!state.selected_template) {
    return {
      templateRel: "",
      message: "QDM_INJECT_TEMPLATE no selectedTemplate in session state. Do not guess a template; continue without template injection.",
    };
  }
  if (!isAllowedTemplateSelectionPath(state.selected_template)) {
    return {
      templateRel: state.selected_template,
      message: `QDM_INJECT_TEMPLATE template must be templates/... or reports/.../template.md: ${state.selected_template}.`,
    };
  }
  let resolver;
  try {
    resolver = newPathResolver(root);
  } catch (error) {
    return {
      templateRel: state.selected_template,
      message: `QDM_INJECT_TEMPLATE path config error: ${error.message || error}.`,
    };
  }
  try {
    const info = statSync(resolver.resolve(state.selected_template));
    if (info.isDirectory()) {
      return { templateRel: state.selected_template, message: `QDM_INJECT_TEMPLATE missing ${state.selected_template}.` };
    }
  } catch {
    return { templateRel: state.selected_template, message: `QDM_INJECT_TEMPLATE missing ${state.selected_template}.` };
  }
  return { templateRel: state.selected_template, message: "" };
}

function stripMarkdownFrontmatter(text) {
  const lines = String(text).split("\n");
  if (!lines.length || lines[0].trim() !== "---") return text;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === "---") return lines.slice(i + 1).join("\n");
  }
  return text;
}

function normalizeCommand(command) {
  return String(command || "").replaceAll("\\", "/").trim().split(/\s+/).filter(Boolean).join(" ");
}

export function getReportState(state, reportName = "template") {
  if (!state.reports) state.reports = {};
  if (!reportName) reportName = "template";
  if (!state.reports[reportName]) {
    state.reports[reportName] = { recorded_modules: [], template_injected: false };
  }
  const current = state.reports[reportName];
  if (!Array.isArray(current.recorded_modules)) current.recorded_modules = [];
  return current;
}

export function addModule(reportState, module) {
  if (reportState.recorded_modules.includes(module)) return false;
  reportState.recorded_modules.push(module);
  return true;
}

function recordTemplateDiagnostic(root, sessionID, session, reportState, outcome, templatePath) {
  if (process.env.QDM_HARNESS_DIAG !== "1") return;
  let resolver;
  try {
    resolver = newPathResolver(root);
  } catch {
    const resourceRoot = normalizeResolverOwners(root).resourceRoot;
    resolver = { resolve: (rel) => path.join(resourceRoot, rel) };
  }
  const event = {
    ts: new Date().toISOString(),
    session_id: sessionID,
    event: "inject_template",
    selected_playbook: session.selected_playbook,
    selected_playbooks: session.selected_playbooks,
    mode: session.mode,
    composite: session.composite,
    template_path: templatePath,
    template_stats: pathStats(resolver.resolve(templatePath || "")),
    template_already_injected: Boolean(reportState?.template_injected),
    outcome,
  };
  try {
  const dir = diagnosticsDir(root);
    mkdirSync(dir, { recursive: true });
    appendFileSync(path.join(dir, `${safeSessionId(sessionID)}.jsonl`), `${JSON.stringify(event)}\n`);
  } catch {
    // diagnostics are best-effort
  }
}

function pathStats(filePath) {
  try {
    const data = readFileSync(filePath);
    return { path: filePath, exists: true, bytes: data.length, lines: countLines(data) };
  } catch {
    return { path: filePath, exists: false, bytes: 0, lines: 0 };
  }
}

function countLines(data) {
  if (!data.length) return 0;
  let lines = 0;
  for (const byte of data) {
    if (byte === 10) lines += 1;
  }
  if (data[data.length - 1] !== 10) lines += 1;
  return lines;
}

function parsePayload(input) {
  try {
    const text = Buffer.isBuffer(input) ? input.toString("utf8") : String(input || "");
    return JSON.parse(text);
  } catch {
    return null;
  }
}

function escapeRegExp(value) {
  return String(value).replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
